reframe/core/systems.py

```diff
@@ -13,6 +13,169 @@
+                    'num_cpus_per_socket', 'num_sockets', 'topology']:
+            if var in processor_info:
+                setattr(self, f'_{var}', processor_info[var])
+
+    @property
+    def info(self):
+        '''All the available information from the configuration.
+
+        :type: :class:`dict`
+        '''
+        return self._info
+
+    @property
+    def arch(self):
+        '''The microarchitecture of the processors.
+
+        :type: :class:`str`
+        '''
+        return self._arch
+
+    @property
+    def num_cpus(self):
+        '''Number of logical CPUs.
+
+        :type: integral
+        '''
+        return self._num_cpus
+
+    @property
+    def num_cpus_per_core(self):
+        '''Number of logical CPUs per core.
+
+        :type: integral
+        '''
+        return self._num_cpus_per_core
+
+    @property
+    def num_cpus_per_socket(self):
+        '''Number of logical CPUs per socket.
+
+        :type: integral
+        '''
+        return self._num_cpus_per_socket
+
+    @property
+    def num_sockets(self):
+        '''Number of sockets.
+
+        :type: integral
+        '''
+        return self._num_sockets
+
+    @property
+    def topology(self):
+        '''Topology of the architecture.
+
+        :type: :class:`Dict[str, obj]`
+        '''
+        return self._topology
+
+    @property
+    def num_cores(self):
+        '''Number of cores. Derived by ``num_cpus // num_cpus_per_core``.
+
+        :type: integral
+        '''
+        if self._num_cpus and self._num_cpus_per_core:
+            return self._num_cpus // self._num_cpus_per_core
+        else:
+            return None
+
+    @property
+    def num_cores_per_socket(self):
+        '''Number of cores per socket. Derived by
+        ``num_cores // num_sockets``.
+
+        :type: integral
+        '''
+        if self.num_cores and self._num_sockets:
+            return self.num_cores // self._num_sockets
+        else:
+            return None
+
+    def __str__(self):
+        return json.dumps(self._info, indent=2)
+
+
+class Device(jsonext.JSONSerializable):
+    '''A representation of a device inside ReFrame.
+
+    .. warning::
+       Users may not create :class:`Device` objects directly.
+    '''
+
+    def __init__(self, device_info):
+        self._type = None
+        self._arch = None
+        self._vendor = None
+        self._info = device_info
+
+        if device_info == {}:
+            return
+
+        for var in ['arch', 'type', 'vendor']:
+            if var in device_info:
+                setattr(self, f'_{var}', device_info[var])
+
+    @property
+    def info(self):
+        '''All the available information from the configuration.
+
+        :type: :class:`dict`
+        '''
+        return self._info
+
+    @property
+    def arch(self):
+        '''The architecture of the device.
+
+        :type: :class:`str`
+        '''
+        return self._arch
+
+    @property
+    def type(self):
```

Comment:
I would call it `device_type`  everywhere, in order not to confuse with Python's `type`.